src/coreclr/jit/promotion.cpp

```diff
@@ -621,6 +621,38 @@ class LocalUses
     bool EvaluateReplacement(
         Compiler* comp, unsigned lclNum, const Access& access, unsigned inducedCount, weight_t inducedCountWtd)
     {
+        // Verify that this replacement has proper GC ness compared to the
+        // layout. While reinterpreting GC fields to integers can be considered
+        // UB, there are scenarios where it can happen safely:
+        //
+        // * The user code could have guarded the access with a dynamic check
+        //   that it doesn't contain a GC pointer, so that the access is actually
+        //   in dead code. This happens e.g. in span functions in SPC.
+        //
+        // * For byrefs, reinterpreting as an integer could be ok in a
+        //   restricted scope due to pinning.
```

Comment:
The essence of this issue is about scope. If you have reachable code which does these things then it would be fine for physical promotion to kick in and promote with the other type. That is no more or less undefined than what already happens in the user code.